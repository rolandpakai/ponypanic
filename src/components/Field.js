import { useMemo, useContext } from "react";

import { ThemeContext } from '../contexts/ThemeContext';
import { randomInteger, validateHeroAction } from '../utils/util';
import { FIELD_TYPE, TREASURE_COUNT, MAP_COUNT, ENEMY_COUNT } from '../utils/constants';

const Field = ( props ) => {
  const { theme } = useContext(ThemeContext);
  const { id, collectedByHeroId, health, name, playerId, position, size, score, action, type, level } = {...props};

  let randomTreasure = useMemo(() => randomInteger(1, TREASURE_COUNT), [level]);
  let randomEnemy = useMemo(() => randomInteger(1, ENEMY_COUNT), [level]);

  const mapLevel = level > MAP_COUNT ? MAP_COUNT : level;

  const divStyle = {
    width: size,
    height: size,
    backgroundSize: size,
    backgroundRepeat: 'no-repeat',
    backgroundImage: `url('./themes/${theme}/maps/map-${mapLevel}/floor.png')`,
  }

  let imgStyle = {
    width: size,
    height: size,
    backgroundRepeat: 'no-repeat',
    src: '',
  };

  switch (type) {
    case FIELD_TYPE.HERO: {
      const heroAction = validateHeroAction(action);
      imgStyle.src = `./themes/${theme}/heroes/${heroAction}.png`;
      break;
    } 
    case FIELD_TYPE.ENEMY: {
      imgStyle.src = `./themes/${theme}/enemies/enemy-${randomEnemy}.png`;
      break;
    }
    case FIELD_TYPE.BULLET: {
      imgStyle.src = `./themes/${theme}/bullets/bullet.png`;
      break;
    }
    case FIELD_TYPE.TREASURE: {
      imgStyle.src = `./themes/${theme}/treasures/treasure-${randomTreasure}.png`;
      break;
    }
    case FIELD_TYPE.OBSTACLE: {
      imgStyle.src = `./themes/${theme}/maps/map-${mapLevel}/block.png`;
      break;
    }
    default: imgStyle.backgroundImage = '';
  }

  return (
      <div style={divStyle}>
        { type !== FIELD_TYPE.FLOOR &&
          <img 
            alt={type}
            className=''
            src={imgStyle.src}
            style={imgStyle}
          />
        }
      </div>
  )
}

export default Field;