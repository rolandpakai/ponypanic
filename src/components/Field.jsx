import React, { Fragment, useContext, useMemo, useState } from "react";
import Box from "@mui/material/Box";
import Popover from "@mui/material/Popover";

import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell, { tableCellClasses } from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TableRow from "@mui/material/TableRow";

import { ThemeContext } from "../contexts/ThemeContext";
import { randomInteger, validateHeroAction } from "../utils/util";
import { FIELD_TYPE, MAP_COUNT, TREASURE_COUNT } from "../utils/constants";

const Field = (props) => {
  const [anchorEl, setAnchorEl] = useState();
  const { theme } = useContext(ThemeContext);
  const { idd, size, action, type, level, data } = { ...props };

  const open = Boolean(anchorEl);

  let hasPopOver = false;

  const randomTreasure = useMemo(() => randomInteger(1, TREASURE_COUNT), []);
  const mapLevel = level > MAP_COUNT ? MAP_COUNT : level;

  const handlePopoverOpen = (event) => {
    setAnchorEl(event.currentTarget);
  };

  const handlePopoverClose = () => {
    setAnchorEl(null);
  };

  const imgStyle = {
    width: size,
    height: size,
    backgroundRepeat: "no-repeat",
    margin: "auto",
    src: "",
  };

  let popoverContent;

  switch (type) {
    case FIELD_TYPE.HERO: {
      const { playerId, health, score } = { ...data };
      const heroAction = validateHeroAction(action).toLowerCase();

      imgStyle.src = `./themes/${theme}/heroes/${heroAction}.png`;
      hasPopOver = true;

      popoverContent = (
        <TableBody>
          <TableRow key={`${idd}-playerId`}>
            <TableCell align="left">Id</TableCell>
            <TableCell align="right">{playerId}</TableCell>
          </TableRow>
          <TableRow key={`${idd}-health`}>
            <TableCell align="left">Health</TableCell>
            <TableCell align="right">
              {health.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
          <TableRow key={`${idd}-score`}>
            <TableCell align="left">Score</TableCell>
            <TableCell align="right">{score}</TableCell>
          </TableRow>
        </TableBody>
      );

      break;
    }
    case FIELD_TYPE.ENEMY:
    case FIELD_TYPE.ENEMY_BULLET: {
      const {
        id,
        moveProbability,
        shootProbability,
        onTouchDamage,
        bulletDamage,
        health,
      } = { ...data };

      imgStyle.src = `./themes/${theme}/enemies/${type.toLowerCase()}-${mapLevel}.png`;
      hasPopOver = true;

      popoverContent = (
        <TableBody>
          <TableRow key={`${idd}-id`}>
            <TableCell align="left">Id</TableCell>
            <TableCell align="right">{id}</TableCell>
          </TableRow>
          <TableRow key={`${idd}-health`}>
            <TableCell align="left">Health</TableCell>
            <TableCell align="right">
              {health.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
          <TableRow key={`${idd}-moveProbability`}>
            <TableCell align="left">Move Probability</TableCell>
            <TableCell align="right">
              {moveProbability.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
          <TableRow key={`${idd}-shootProbability`}>
            <TableCell align="left">Shoot Probability</TableCell>
            <TableCell align="right">
              {shootProbability.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
          <TableRow key={`${idd}-bulletDamage`}>
            <TableCell align="left">Bullet Damage</TableCell>
            <TableCell align="right">
              {bulletDamage.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
          <TableRow key={`${idd}-onTouchDamage`}>
            <TableCell align="left">Touch Damage</TableCell>
            <TableCell align="right">
              {onTouchDamage.toLocaleString("en", { style: "percent" })}
            </TableCell>
          </TableRow>
        </TableBody>
      );

      break;
    }
    case FIELD_TYPE.BULLET: {
      // const { id, damage, direction, shotByEnemyId } = { ...data };

      imgStyle.src = `./themes/${theme}/bullets/bullet.png`;
      imgStyle.width = "auto";
      imgStyle.height = "auto";
      break;
    }
    case FIELD_TYPE.TREASURE: {
      const { name } = { ...data };

      imgStyle.src = `./themes/${theme}/treasures/treasure-${randomTreasure}.png`;
      hasPopOver = true;

      popoverContent = (
        <TableBody>
          <TableRow key={`${idd}-name`}>
            <TableCell align="center">{name}</TableCell>
          </TableRow>
        </TableBody>
      );

      break;
    }
    case FIELD_TYPE.OBSTACLE: {
      imgStyle.src = `./themes/${theme}/maps/map-${mapLevel}/block.png`;
      break;
    }
    default:
      imgStyle.backgroundImage = "";
  }

  return (
    <>
      <Box
        aria-owns={hasPopOver && open ? "mouse-over-popover" : undefined}
        aria-haspopup="true"
        onMouseEnter={hasPopOver ? handlePopoverOpen : undefined}
        onMouseLeave={hasPopOver ? handlePopoverClose : undefined}
        sx={{
          width: size,
          height: size,
          display: "flex",
          backgroundSize: size,
          backgroundRepeat: "no-repeat",
          backgroundImage: `url('./themes/${theme}/maps/map-${mapLevel}/floor.png')`,
        }}
      >
        {type !== FIELD_TYPE.FLOOR && (
          <img alt={type} className="" src={imgStyle.src} style={imgStyle} />
        )}
      </Box>
      {hasPopOver && (
        <Popover
          anchorEl={anchorEl}
          anchorOrigin={{
            vertical: "bottom",
            horizontal: "left",
          }}
          disableRestoreFocus
          id="mouse-over-popover"
          onClose={handlePopoverClose}
          open={open}
          sx={{
            pointerEvents: "none",
          }}
          transformOrigin={{
            vertical: "top",
            horizontal: "left",
          }}
        >
          <TableContainer>
            <Table
              aria-label="simple table"
              size="small"
              sx={{
                [`& .${tableCellClasses.root}`]: {
                  borderBottom: "none",
                },
              }}
            >
              <TableHead>
                <TableRow>
                  <TableCell align="center" colSpan={2}>
                    {type}
                  </TableCell>
                </TableRow>
              </TableHead>
              {popoverContent}
            </Table>
          </TableContainer>
        </Popover>
      )}
    </>
  );
};

export default Field;
