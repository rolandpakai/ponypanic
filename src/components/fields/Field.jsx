class Field {
  constructor(idd, type, level, theme, data) {
    this.idd = idd;
    this.level = level;
    this.type = type;
    this.theme = theme;
    this.data = data;

    this.popover = false;
  }

  getImgStyleSrc() {
    return this.imgStyleSrc;
  }

  getPopoverContent() {
    return this.popoverContent;
  }

  hasPopover() {
    return this.popover;
  }
}

export default Field;
