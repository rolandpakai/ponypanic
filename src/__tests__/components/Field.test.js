import { render, act } from '@testing-library/react';
import Field from '../../components/Field';

describe('Field Component', () => {
  test('Field renders', () => {
    const {container} = render(<Field />);

		expect(container).toMatchSnapshot();
  });
});