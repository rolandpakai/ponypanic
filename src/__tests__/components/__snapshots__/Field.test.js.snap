// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Component Field renders 1`] = `
<div>
  <div
    class="MuiBox-root css-nza37f"
  >
    <img
      class=""
      src=""
      style="background-repeat: no-repeat; margin: auto;"
    />
  </div>
</div>
`;
